streams/shared/src/main/scala/zio/stream/experimental/ZStream.scala

```diff
@@ -82,6 +82,73 @@ abstract class ZStream[-R, +E, +O](
   ): ZStream[R1, E1, O1] =
     ZStream.absolve(ev(self))
 
+  /**
+   * Aggregates elements of this stream using the provided sink for as long
+   * as the downstream operators on the stream are busy.
+   *
+   * This operator divides the stream into two asynchronous "islands". Operators upstream
+   * of this operator run on one fiber, while downstream operators run on another. Whenever
+   * the downstream fiber is busy processing elements, the upstream fiber will feed elements
+   * into the sink until it signals completion.
+   *
+   * Any sink can be used here, but see [[Sink.foldWeightedM]] and [[Sink.foldUntilM]] for
+   * sinks that cover the common usecases.
+   */
+  final def aggregateAsync[R1 <: R, E1 >: E, O1 >: O, P](transducer: ZTransducer[R1, E1, O1, P]): ZStream[R1, E1, P] = {
```

Comment:
So I think this implementation needs a bit more work to achieve the previous guarantees. Here are the main problem points:
1. The consumer will keep spinning and taking data from the bucket even if the producer hasn't produced anything. You basically have to take the mechanism you created with `start` and thread it throughout the whole process. This can be done by the producer storing a `Promise` inside the `Ref` that the consumer can await to know when there's data available in the bucket.
2. The producer has no backpressuring mechanism and will overwrite data that hasn't been taken by the consumer (line 124). This can be solved by attaching a `Promise` to the bucket that will be resolved by the consumer when it  took the produced result.
3. I actually think the Semaphore is now redundant because you can do all the bucket swaps atomically.
4. Looking at your implementation, it may be worth exploring the queue option! I think the current encoding removes some of the constraints the previous implementation had.

Reply:
🤦 I don't know why I keep thinking there is going to be a strict order of takes. Will address this.

Reply:
I think a bounded queue of 1 is exactly the semantic I'm looking for. Like an `IVar` that can be emptied and filled multiple times.